### PR TITLE
Elan and Text file processing from datasets (WIP)

### PR DESCRIPTION
**Pull request by harrykeightley**
- Previous PR confirmed that the processing "flow" worked correctly without the processing actually happening (it was just copying files across to the new bucket and then marking the dataset as processed).
- This is a cautious attempt to do file processing for transcription files.
- Elan file tier information is extracted and then cleaned
- Text file transcription info is just cleaned and converted to json.

### CODE REVIEW

File: api/functions/process_file.py [hunk 1]
```diff
@@ -44,36 +47,128 @@ def process_dataset_file(event: pubsub_v1.types.message, context: Context) -> No
...
     else:
         raise RuntimeError("Unrecognised file extension. Processing halted.")
 
     # Save the processed file to the datasets folder in cloud storage
     datasets_bucket_name = os.environ.get("USER_DATASETS_BUCKET")
-    upload_blob(datasets_bucket_name, local_file, f"{uid}/{dataset_name}/{file_name}")
+    upload_blob(
+        datasets_bucket_name,
+        processed_file,
+        f"{uid}/{dataset_name}/{processed_file.name}",
```
Comment (nicklambourne) [inline api/functions/process_file.py:70]: nit: kwarg use could make this more readable

File: api/functions/process_file.py [hunk 2]
```diff
@@ -44,36 +47,128 @@ def process_dataset_file(event: pubsub_v1.types.message, context: Context) -> No
...
+        punctuation_to_explode_by=options["punctuationToReplace"],
+        special_cases=options["wordsToRemove"].split("\n"),
+        translation_tags=options["tagsToRemove"].split("\n"),
+    )
 
 
-def process_audio_file(file_name: str) -> None:
-    pass
+def process_audio_file(file: Path) -> Path:
+    return file
```
Comment (nicklambourne) [inline api/functions/process_file.py:171]: Should there be a `#TODO:` here?

File: api/functions/requirements.txt [hunk 3]
```diff
@@ -2,4 +2,5 @@ google-cloud-storage
 google-cloud-pubsub
 google-auth
 firebase-admin
-functions-framework==3.*
\ No newline at end of file
+functions-framework==3.*
```
Comment (nicklambourne) [inline api/functions/requirements.txt:5]: Any reason not to pin these at a specifc version? Upstream changes can be a nightmare (and hard to notice, especially in a lambda-like environment)

File: api/functions/utils/clean_json.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+    json_data: List[Dict[str, str]],
+    punctuation_to_collapse_by: str = "",
+    punctuation_to_explode_by: str = "",
+    special_cases: set = {None},
+    translation_tags: set = {None},
+    remove_english: bool = False,
+) -> List[Dict[str, str]]:
+    """
+    Clean a list of utterances (Python dictionaries) based on the given parameters.
+    :param json_data: List of Python dictionaries, each must have a 'transcript' key-value.
```
Comment (nicklambourne) [inline api/functions/utils/clean_json.py:29]: Nit: please add a TODO here to google-style these.

File: api/functions/utils/clean_json.py [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+            text=word,
+            punctuation_to_collapse_by=punctuation_to_collapse_by,
+            punctuation_to_explode_by=punctuation_to_explode_by,
+        )
+        clean_words.append(word)
+    return clean_words, english_word_count
+
+
+def get_english_words() -> Set[str]:
+    """ """
```
Comment (nicklambourne) [inline api/functions/utils/clean_json.py:148]: lol wat 😂

**Review by nicklambourne — APPROVED** [review]
Couple of questions/notes/nits. LGTM even if our old scripts are a little on yikes side 😂.

File: api/functions/process_file.py [hunk 6]
```diff
@@ -44,36 +47,128 @@ def process_dataset_file(event: pubsub_v1.types.message, context: Context) -> No
...
+
+
+def clean_data(
+    transcription_data: List[Dict[str, str]], options: Dict[str, Any]
+) -> List[Dict[str, str]]:
+    """Takes a list of utterance information and cleans it based off the provided
+    dataset processing options.
+
+    Parameters:
+        transcription_data: A list of information about utternaces in the transcription.
```
Comment (aviraljain99) [inline api/functions/process_file.py:155]: `utternances` - typo, should be `utterances`.

**Review by aviraljain99 — APPROVED** [review]
Looks good to me! Had a small discussion with @harrykeightley about the procedural nature of this code and could maybe it could use some classes? Something that uses classes could even be easier to test - but I am not 100% sure. What do you think @nicklambourne. I am happy with how this is tho atm!

File: api/functions/utils/elan_to_json.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    :param corpus_tiers_file list of all
+    :return: a list of dictionaries, where each dictionary is an annotation
+    """
+
+    print(
+        f"processing eaf {input_elan_file} using {tier_order} {tier_type} {tier_name}"
+    )
+
+    _, full_file_name = os.path.split(input_elan_file)
+    file_name, _ = os.path.splitext(full_file_name)
```
[the remark this answers is not in the document]
Comment (nicklambourne) [reply api/functions/utils/elan_to_json.py:43]: Pathlib should be enforced imo.

**Comment by nicklambourne** [comment]
> Looks good to me! Had a small discussion with @harrykeightley about the procedural nature of this code and could maybe it could use some classes? Something that uses classes could even be easier to test - but I am not 100% sure. What do you think @nicklambourne. I am happy with how this is tho atm!

This is fairly straightforward batch data processing, so I'm sceptical of the need for classes at all wrt the top-level organisation of the functions. I'm open to concrete suggestions, but in my experience there's not much easier to (unit) test than plain old functions with no side effects.The only place I can see a potential application for it might be as dataclasses encapsulating e.g. utterances or pre-EAF data.

If I was going to improve it, my focus would be on the following:
- Smaller functions
- _Significantly less_ parameters per function
- Removing questionable default parameters (`""`, `{None}`)
- Replacement of prints with calls to logging
- Actioning or removing TODOs
- Simplifying the "clean" function logic

But given it "works" (in the loosest sense of the word) and there's so much work still to do I wouldn't blame you for leaving the cleanup until later.

File: api/functions/process_file.py (hunk 1 above)
Comment (harrykeightley) [reply api/functions/process_file.py]: Fixed in next revision

File: api/functions/process_file.py (hunk 2 above)
Comment (harrykeightley) [reply api/functions/process_file.py:175]: Fixed in next revision

File: api/functions/requirements.txt (hunk 3 above)
Comment (harrykeightley) [reply api/functions/requirements.txt]: I actually added this back in for no reason- we're not using this as of now so I've taken it out but have seen that the rest of our requirements.txt file is full of unpinned dependencies, so I'm making an issue to remedy this but not addressing it in this PR.

File: api/functions/utils/clean_json.py (hunk 4 above)
Comment (harrykeightley) [reply api/functions/utils/clean_json.py:29]: Fixed in next revision

File: api/functions/utils/clean_json.py (hunk 5 above)
Comment (harrykeightley) [reply api/functions/utils/clean_json.py]: Oh no, I've been caught!
- Previously this was a call which would download a set of english words from an online source.
- We weren't currently using this feature in the processing options, and I didn't want to have to add an unused dependency at this stage so I decided to gimp the method.
- My thinking is that if the old scripts work, then we probably have better things to be directing efforts to at this stage. If instead, we have issues, I'm seriously considering a refactor, which will include separating out/removing the english word filtering.

File: api/functions/process_file.py (hunk 6 above)
Comment (harrykeightley) [reply api/functions/process_file.py]: Good catch. Fixed in next revision

File: api/functions/utils/elan_to_json.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (harrykeightley) [reply api/functions/utils/elan_to_json.py]: Fixed in next revision

**Comment by harrykeightley** [comment]
Merging after fixing the majority of raised nits.

**Pull request merged**